TutorialsAndTests/Tests/AtlHenTests.cpp

```diff
@@ -81,22 +84,30 @@ TEST(AtlHenTests, RequireThat_Cluck_IsExecutedOnMainThread_WhenCalledFromWorkerT
             return S_OK; // Feel free to put a breakpoint here, and see which thread we are called on
         }));
 
-    CComPtr<IHen> hen;
-    HR(CoCreateInstance(CLSID_AtlHen, nullptr, CLSCTX_INPROC_SERVER, IID_IHen, reinterpret_cast<void**>(&hen)));
+    ComPtr<IHen> hen;
+    HR(CoCreateInstance(CLSID_AtlHen, nullptr, CLSCTX_INPROC_SERVER, IID_IHen, &hen));
 
     // To play nice with COM, we should not pass COM interface pointers to other threads
     // unless they are 'agile'. We can do this with a simple wrapper that converts the interfaces
     // to agile interfaces. We could also marshal the interface to a stream using CoMarshalInterThreadInterfaceInStream
-    // and CoGetInterfaceAndReleaseStream.
-    const auto agileHen = AgilePtr<IHen>(hen);
+    // and CoGetInterfaceAndReleaseStream. Also use AgileRef from WRL to show how it can be used for marshaling.
+    AgileRef agileHen;
+    // AsAgile also use RoGetAgileReference.
+    HR(hen.AsAgile(&agileHen));
     const auto agileObserver = AgilePtr<IAsyncCluckObserver>(observer);
 
     std::thread thread{ [agileHen, agileObserver, mainThreadId]
     {
         ComRuntime runtime{Apartment::MultiThreaded};
 
         SetThreadDescription(GetCurrentThread(), L"Worker thread"); // Help debugging
-        HR(agileHen.Get()->CluckAsync(agileObserver.Get()));
+        ComPtr<IHen> henLocal;
+        // All methods in AgileRef calls Resolve to get pointer what is safe to use in this thread.
+        HR(agileHen.As(&henLocal));
+        HR(henLocal->CluckAsync(agileObserver.Get()));
+        // Manually reset ComPtr to avoid deadlock related to destruction after
+        // PostThreadMessage(mainThreadId, WM_QUIT, 0, 0);
+        henLocal.Reset();
```

Comment:
Could you try to figure out why the main thread apartment has to be pumping when henLocal is destroyed? I suspect that henLocal is a proxy, and that when the refcount goes to 0, a message is sent to the main thread to destroy the stub. It should be straight forward to put a breakpoint in the hen destructor to see that it is indeed called from within the message pump? If this is the case, you can maybe make the docstring above explain why we get a deadlock?

Reply:
Yes, it is a destruction of proxy object. I took for granted COM rules, and did paid attention that such string might be helpful.
I will change my PR.